gtsam/discrete/DiscreteFactorGraph.cpp

```diff
@@ -121,6 +121,11 @@ namespace gtsam {
     for (auto&& factor : factors) product = (*factor) * product;
     gttoc(product);
 
+    // Normalize the product factor to prevent underflow
+    Ordering ordering = Ordering::Colamd(DiscreteFactorGraph{product});
```

Comment:
Can’t we get keys directly?

Reply:
Yes, I don’t see why not 👍

Reply:
I understand the code now. I tried on my machine, and the following works (and is more efficient):
```c++
    // Sum all potentials by pretending all keys are frontal:
    auto normalization = product.sum(product.size());

    // Normalize the product factor to prevent underflow.
    product = product / *normalization;
```